### DISCUSSION TITLE
Failed to load Python DLL - seems something's going wrong during the build process

### DISCUSSION THREAD

**Discussion by AnAnalogGuy** [pyinstaller/pyinstaller#8656] [Help] [answered]
I created a little Systray App (HueConnect) in Python. The script runs fine from the command line. 

When building in onedir-mode, the build process with PyInstaller runs successfully (no errors shown and process completed successfully). When starting the application from /build/HueConnect, a popup error is shown and the application does not start up:

![grafik](https://github.com/user-attachments/assets/7813bae9-d6b3-4c67-aec4-dc4d93085ddc)

`_(the specified module could not be found)_`

I checked Analysis-00.toc and EXE-00.toc and both contain references for the related python311.dll pointing to the correct location:

`('python311.dll', 'C:\\Program Files\\Python311\\python311.dll', 'BINARY'),`

Interestingly in the error message the path to python311.dll is different: `J:\hueconnect\build\HueConnect\_internal\python311.dll`. 

While  J:\hueconnect\build\HueConnect exists, there is no "_internal" directory below that path.

After creating the "_internal" directory manually, copying python311.dll to it and trying to start the application again, a different Error pops up:

![grafik](https://github.com/user-attachments/assets/8e98b49f-121f-4fb5-9995-8e396acc5c1e)

Here I'm stuck: Obv. something with embedding the pyhton interpreter and needed runtime libs did not work while building the EXE with pyinstaller. However, there is no errors issued when building the EXE:

_J:\hueconnect>pyinstaller --onedir --clean --name HueConnect --add-data "hueconnect.conf;." --add-binary "icon_default.jpg;." --noconsole --noconfirm .\HueConnect.py
990 INFO: PyInstaller: 6.9.0, contrib hooks: 2024.7
990 INFO: Python: 3.11.3
1000 INFO: Platform: Windows-10-10.0.19045-SP0
1001 INFO: Python environment: J:\hueconnect\.venv
1006 INFO: wrote J:\hueconnect\HueConnect.spec
1034 INFO: Removing temporary files and cleaning cache in C:\Users\Frank\AppData\Local\pyinstaller
1049 INFO: Module search paths (PYTHONPATH):
['J:\\hueconnect\\.venv\\Scripts\\pyinstaller.exe',
 'C:\\Program Files\\Python311\\python311.zip',
 'C:\\Program Files\\Python311\\DLLs',
 'C:\\Program Files\\Python311\\Lib',
 'C:\\Program Files\\Python311',
 'J:\\hueconnect\\.venv',
 'J:\\hueconnect\\.venv\\Lib\\site-packages',
 'J:\\hueconnect']
1463 INFO: Appending 'binaries' from .spec
1463 INFO: Appending 'datas' from .spec
1464 INFO: checking Analysis
1465 INFO: Building Analysis because Analysis-00.toc is non existent
1465 INFO: Running Analysis Analysis-00.toc
1465 INFO: Target bytecode optimization level: 0
1465 INFO: Initializing module dependency graph...
1469 INFO: Caching module graph hooks...
1579 INFO: Analyzing base_library.zip ...
3621 INFO: Loading module hook 'hook-heapq.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
3724 INFO: Loading module hook 'hook-encodings.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
5764 INFO: Loading module hook 'hook-pickle.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
8134 INFO: Caching module dependency graph...
8257 INFO: Looking for Python shared library...
8264 INFO: Using Python shared library: C:\Program Files\Python311\python311.dll
8264 INFO: Analyzing J:\hueconnect\HueConnect.py
8379 INFO: Loading module hook 'hook-PIL.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
8553 INFO: Loading module hook 'hook-PIL.Image.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
9950 INFO: Loading module hook 'hook-packaging.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
10063 INFO: Loading module hook 'hook-PIL.ImageFilter.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
11559 INFO: Loading module hook 'hook-charset_normalizer.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
11868 INFO: Loading module hook 'hook-certifi.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
12162 INFO: Loading module hook 'hook-pystray.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
13002 INFO: Processing pre-safe import module hook six.moves from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-six.moves.py'.
13183 INFO: Processing pre-safe import module hook gi from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.py'.
13383 INFO: Processing module hooks...
13481 INFO: Loading module hook 'hook-pkg_resources.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
15551 INFO: Loading module hook 'hook-platform.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
15603 INFO: Loading module hook 'hook-xml.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
15656 INFO: Loading module hook 'hook-sysconfig.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
16266 INFO: Loading module hook 'hook-PIL.SpiderImagePlugin.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
17334 INFO: Performing binary vs. data reclassification (6 entries)
17407 INFO: Looking for ctypes DLLs
17455 INFO: Analyzing run-time hooks ...
17457 INFO: Including run-time hook 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_inspect.py'
17468 INFO: Including run-time hook 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_pkgres.py'
17479 INFO: Including run-time hook 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_pkgutil.py'
17486 INFO: Processing pre-find module path hook _pyi_rth_utils from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\pre_find_module_path\\hook-_pyi_rth_utils.py'.
17500 INFO: Loading module hook 'hook-_pyi_rth_utils.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
17511 INFO: Looking for dynamic libraries
18363 INFO: Extra DLL search directories (AddDllDirectory): []
18363 INFO: Extra DLL search directories (PATH): []
18842 INFO: Warnings written to J:\hueconnect\build\HueConnect\warn-HueConnect.txt
18948 INFO: Graph cross-reference written to J:\hueconnect\build\HueConnect\xref-HueConnect.html
19144 INFO: checking PYZ
19144 INFO: Building PYZ because PYZ-00.toc is non existent
19145 INFO: Building PYZ (ZlibArchive) J:\hueconnect\build\HueConnect\PYZ-00.pyz
19824 INFO: Building PYZ (ZlibArchive) J:\hueconnect\build\HueConnect\PYZ-00.pyz completed successfully.
19847 INFO: checking PKG
19848 INFO: Building PKG because PKG-00.toc is non existent
19849 INFO: Building PKG (CArchive) HueConnect.pkg
20052 INFO: Building PKG (CArchive) HueConnect.pkg completed successfully.
20055 INFO: Bootloader J:\hueconnect\.venv\Lib\site-packages\PyInstaller\bootloader\Windows-64bit-intel\runw.exe
20055 INFO: checking EXE
20056 INFO: Building EXE because EXE-00.toc is non existent
20056 INFO: Building EXE from EXE-00.toc
20059 INFO: Copying bootloader EXE to J:\hueconnect\build\HueConnect\HueConnect.exe
20182 INFO: Copying icon to EXE
20268 INFO: Copying 0 resources to EXE
20268 INFO: Embedding manifest in EXE
20362 INFO: Appending PKG archive to EXE
20519 INFO: Fixing EXE headers
21032 INFO: Building EXE from EXE-00.toc completed successfully.
21035 INFO: checking COLLECT
21036 INFO: Building COLLECT because COLLECT-00.toc is non existent
21037 INFO: Building COLLECT COLLECT-00.toc
22129 INFO: Building COLLECT COLLECT-00.toc completed successfully._

Does anyone has an idea would could have gone wrong/what i might have missded? Any idea/hint is very welcome. Thanks a lot!

**Comment by rokm** [comment]
> ```J:\hueconnect\build\HueConnect\_internal\python311.dll.```

You seem to be running .exe in the `build` directory, which contains intermediate build results. The final application bundle is in the `dist` directory - i.e., `J:\hueconnect\dist\HueConnect`. There, you should find the `_internal` directory next to the executable, with all required files in it.

**Reply by AnAnalogGuy** [reply]
Wow, you're fast :-)

You're right, i was running the application from the build directory. I should have mentioned: When running J:\hueconnect\dist\HueConnect from cmd, the application doesn't start. Only thing that happens is a line feed and the application returns without any output. 

![grafik](https://github.com/user-attachments/assets/21812f38-6d7c-41d2-8603-79cd4eedbd0e)

That's why i tried to run it from /build to get some more "insight".

**Reply by rokm** [reply]
What if you add `--debug bootloader` to your PyInstaller command? Do you get any output from debug-enabled executable?

**Reply by rokm** [reply]
Ah, actually, nevermind. I notice you are using `--noconsole` - so your application does use console for its output.

**Reply by rokm** [reply]
Try removing `--noconsole` to get console output, and once you have that running as expected, try to make a windowed/noconsole build.

**Reply by AnAnalogGuy** [reply]
Update: It's a windows file lock from a backup process that just started. I forgot it's running right now. However, as it just started, it think it's unrelated to the intial problem. I will end the process and go on with the --nosonsole approach. 

--

Now that's interesting: With `-debug bootloader` an error is thrown during build (access permissions to the dist directory). This error hasn't been shown before (without the debug option). I will have a closer look to it, to make sure it's not a backup or other process blocking access, as i would expect this error also would have been thrown earlier without the debug option. 

J:\hueconnect>pyinstaller --onedir --clean --name HueConnect --add-data "hueconnect.conf;." --add-binary "icon_default.jpg;." --noconsole --noconfirm --debug bootloader .\HueConnect.py
898 INFO: PyInstaller: 6.9.0, contrib hooks: 2024.7
898 INFO: Python: 3.11.3
907 INFO: Platform: Windows-10-10.0.19045-SP0
907 INFO: Python environment: J:\hueconnect\.venv
913 INFO: wrote J:\hueconnect\HueConnect.spec
927 INFO: Removing temporary files and cleaning cache in C:\Users\Frank\AppData\Local\pyinstaller
1045 INFO: Module search paths (PYTHONPATH):
['J:\\hueconnect\\.venv\\Scripts\\pyinstaller.exe',
 'C:\\Program Files\\Python311\\python311.zip',
 'C:\\Program Files\\Python311\\DLLs',
 'C:\\Program Files\\Python311\\Lib',
 'C:\\Program Files\\Python311',
 'J:\\hueconnect\\.venv',
 'J:\\hueconnect\\.venv\\Lib\\site-packages',
 'J:\\hueconnect']
1415 INFO: Appending 'binaries' from .spec
1416 INFO: Appending 'datas' from .spec
1417 INFO: checking Analysis
1419 INFO: Building Analysis because Analysis-00.toc is non existent
1419 INFO: Running Analysis Analysis-00.toc
1419 INFO: Target bytecode optimization level: 0
1420 INFO: Initializing module dependency graph...
1421 INFO: Caching module graph hooks...
1515 INFO: Analyzing base_library.zip ...
3516 INFO: Loading module hook 'hook-heapq.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
3588 INFO: Loading module hook 'hook-encodings.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
5404 INFO: Loading module hook 'hook-pickle.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
6513 INFO: Caching module dependency graph...
6634 INFO: Looking for Python shared library...
6644 INFO: Using Python shared library: C:\Program Files\Python311\python311.dll
6644 INFO: Analyzing J:\hueconnect\HueConnect.py
6729 INFO: Loading module hook 'hook-PIL.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
6842 INFO: Loading module hook 'hook-PIL.Image.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
8136 INFO: Loading module hook 'hook-packaging.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
8238 INFO: Loading module hook 'hook-PIL.ImageFilter.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
9592 INFO: Loading module hook 'hook-charset_normalizer.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
9838 INFO: Loading module hook 'hook-certifi.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
10093 INFO: Loading module hook 'hook-pystray.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
10940 INFO: Processing pre-safe import module hook six.moves from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-six.moves.py'.
11128 INFO: Processing pre-safe import module hook gi from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-gi.py'.
11301 INFO: Processing module hooks...
11395 INFO: Loading module hook 'hook-pkg_resources.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
13395 INFO: Loading module hook 'hook-platform.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
13438 INFO: Loading module hook 'hook-xml.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
13500 INFO: Loading module hook 'hook-sysconfig.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
14046 INFO: Loading module hook 'hook-PIL.SpiderImagePlugin.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
14952 INFO: Performing binary vs. data reclassification (6 entries)
15016 INFO: Looking for ctypes DLLs
15048 INFO: Analyzing run-time hooks ...
15051 INFO: Including run-time hook 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_inspect.py'
15062 INFO: Including run-time hook 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_pkgres.py'
15081 INFO: Including run-time hook 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_pkgutil.py'
15088 INFO: Processing pre-find module path hook _pyi_rth_utils from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks\\pre_find_module_path\\hook-_pyi_rth_utils.py'.
15101 INFO: Loading module hook 'hook-_pyi_rth_utils.py' from 'J:\\hueconnect\\.venv\\Lib\\site-packages\\PyInstaller\\hooks'...
15111 INFO: Looking for dynamic libraries
16147 INFO: Extra DLL search directories (AddDllDirectory): []
16147 INFO: Extra DLL search directories (PATH): []
16616 INFO: Warnings written to J:\hueconnect\build\HueConnect\warn-HueConnect.txt
16750 INFO: Graph cross-reference written to J:\hueconnect\build\HueConnect\xref-HueConnect.html
16928 INFO: checking PYZ
16929 INFO: Building PYZ because PYZ-00.toc is non existent
16929 INFO: Building PYZ (ZlibArchive) J:\hueconnect\build\HueConnect\PYZ-00.pyz
17641 INFO: Building PYZ (ZlibArchive) J:\hueconnect\build\HueConnect\PYZ-00.pyz completed successfully.
17662 INFO: checking PKG
17662 INFO: Building PKG because PKG-00.toc is non existent
17662 INFO: Building PKG (CArchive) HueConnect.pkg
17866 INFO: Building PKG (CArchive) HueConnect.pkg completed successfully.
17871 INFO: Bootloader J:\hueconnect\.venv\Lib\site-packages\PyInstaller\bootloader\Windows-64bit-intel\runw_d.exe
17871 INFO: checking EXE
17872 INFO: Building EXE because EXE-00.toc is non existent
17872 INFO: Building EXE from EXE-00.toc
17875 INFO: Copying bootloader EXE to J:\hueconnect\build\HueConnect\HueConnect.exe
18060 INFO: Copying icon to EXE
18199 INFO: Copying 0 resources to EXE
18199 INFO: Embedding manifest in EXE
18314 INFO: Appending PKG archive to EXE
18477 INFO: Fixing EXE headers
18993 INFO: Building EXE from EXE-00.toc completed successfully.
18998 INFO: checking COLLECT
18998 INFO: Building COLLECT because COLLECT-00.toc is non existent
19000 INFO: Removing dir J:\hueconnect\dist\HueConnect
Traceback (most recent call last):
  File "J:\hueconnect\.venv\Lib\site-packages\PyInstaller\building\utils.py", line 381, in _make_clean_directory
    os.remove(path)
PermissionError: [WinError 5] Zugriff verweigert: 'J:\\hueconnect\\dist\\HueConnect'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "J:\hueconnect\.venv\Scripts\pyinstaller.exe\__main__.py", line 7, in <module>
  File "J:\hueconnect\.venv\Lib\site-packages\PyInstaller\__main__.py", line 231, in _console_script_run
    run()
  File "J:\hueconnect\.venv\Lib\site-packages\PyInstaller\__main__.py", line 215, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "J:\hueconnect\.venv\Lib\site-packages\PyInstaller\__main__.py", line 70, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "J:\hueconnect\.venv\Lib\site-packages\PyInstaller\building\build_main.py", line 1216, in main
    build(specfile, distpath, workpath, clean_build)
  File "J:\hueconnect\.venv\Lib\site-packages\PyInstaller\building\build_main.py", line 1156, in build
    exec(code, spec_namespace)
  File "J:\hueconnect\HueConnect.spec", line 36, in <module>
    coll = COLLECT(
           ^^^^^^^^
  File "J:\hueconnect\.venv\Lib\site-packages\PyInstaller\building\api.py", line 1117, in __init__
    self.__postinit__()
  File "J:\hueconnect\.venv\Lib\site-packages\PyInstaller\building\datastruct.py", line 184, in __postinit__
    self.assemble()
  File "J:\hueconnect\.venv\Lib\site-packages\PyInstaller\building\api.py", line 1129, in assemble
    _make_clean_directory(self.name)
  File "J:\hueconnect\.venv\Lib\site-packages\PyInstaller\building\utils.py", line 383, in _make_clean_directory
    _rmtree(path)
  File "J:\hueconnect\.venv\Lib\site-packages\PyInstaller\building\utils.py", line 408, in _rmtree
    shutil.rmtree(path)
  File "C:\Program Files\Python311\Lib\shutil.py", line 759, in rmtree
    return _rmtree_unsafe(path, onerror)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\shutil.py", line 626, in _rmtree_unsafe
    onerror(os.rmdir, path, sys.exc_info())
  File "C:\Program Files\Python311\Lib\shutil.py", line 624, in _rmtree_unsafe
    os.rmdir(path)
OSError: [WinError 145] Das Verzeichnis ist nicht leer: 'J:\\hueconnect\\dist\\HueConnect'

**Reply by rokm** [reply]
My first guess would actually be that your windowed/noconsole appliation is running in the background, and that's why you cannot remove its contents. Look at the system tray (since you say it's a systray app), or open Task Manager and find the running process.

**Reply by AnAnalogGuy** [reply]
The systray and the task manager were the first things i checked :-) There was no related process. 
The lock is removed in the meantime.

**Reply by rokm** [reply, marked as answer]
Well, as I wrote earlier, just adding `--debug` won't do anything, because you are also using `--noconsole` (which I missed initially). So you'll have to also remove `--noconsole` if you want any console output at all.

**Reply by AnAnalogGuy** [reply]
Yes, I understood, you're directions were very clear. I just followed your advice (recompiling without --noconsole)  and i got an error message from my application on the console - it can't resolve the relative path to the config file. My bad :-( So PyInstaller was working perfectly!

Thanks a lot Rok for your fast, friendly and professional support. I'm a bit embarrassed i didn't think of removing --noconsole myself - noob mistake :-)

--
Now it's running - a little small icon with a context menu to control Philips Hue Lightstrips.
![grafik](https://github.com/user-attachments/assets/d978e10a-5d11-44ba-87bf-fa70199b5d50)
